src/mame/altos/altos586_hdc.cpp

```diff
@@ -0,0 +1,371 @@
+		LOG("WRITE FORMAT command\n");
+		if (m_secoffset != 1)
+			logerror("expected one value in data buffer, has %d\n", m_secoffset);
+
+		m_secoffset = 1; // Leave the sector number in.
+		m_iop->drq1_w(ASSERT_LINE);
+		break;
+
+	case 0x10:
+		// Seek to cylinder.
+		// Current cylinder is in data buffer, new one in a separate latch.
+		LOG("SEEK from cylinder %d to %d\n", m_sector[1] << 8 | m_sector[0], m_cyl_latch);
+		if (m_secoffset != 2)
+			logerror("expected two values in data buffer, has %d\n", m_secoffset);
+
+		m_secoffset = 0;
+		m_cyl[m_drive] = m_cyl_latch;
+		// Seek done.
+		m_seek_status |= 0x02;
+		// Not busy.
+		m_status &= ~0x01;
+		break;
+
+	case 0x20:
+		// Not sure, actually.
+		LOG("SELECT command\n");
+		// Head/drive selected?
+		m_seek_status |= 0x01;
+		break;
+
+	case 0x80:
+		// This looks like a reset of some sort.
+		LOG("RESET command\n");
+		device_reset();
+		break;
+
+	default:
+		logerror("unknown command 0x%02x\n", data);
+	}
+}
+
+void altos586_hdc_device::altos586_hdc_io(address_map &map)
+{
+	map(0x0000, 0x3fff).ram(); // 16K SRAM.
+	map(0xffd0, 0xffd1).rw(FUNC(altos586_hdc_device::data_r), FUNC(altos586_hdc_device::data_w));
+	map(0xffd2, 0xffd3).w(FUNC(altos586_hdc_device::head_select_w));
+	map(0xffd4, 0xffd5).rw(FUNC(altos586_hdc_device::seek_status_r), FUNC(altos586_hdc_device::cyl_w));
+	map(0xffd6, 0xffd7).rw(FUNC(altos586_hdc_device::status_r), FUNC(altos586_hdc_device::command_w));
+	// Diags write 0x10 to 0xfff8 and XENIX writes 0x82 to 0xffe6.
+	// Not sure what either does.
+	map(0xfff8, 0xfff9).nopw();
+}
+
+std::error_condition altos586_hdc_device::hdd_load(device_image_interface &image, uint8_t index)
+{
+	if (m_hdd[index]->get_info().sectorbytes != 512) {
+		logerror("expected 512 bytes per sector, got %d", m_geom[index]->sectorbytes);
+		return image_error::INVALIDLENGTH;
+	}
+
+	m_geom[index] = &m_hdd[index]->get_info();
+	return std::error_condition();
+}
+
+void altos586_hdc_device::device_add_mconfig(machine_config &config)
+{
+	I8089(config, m_iop, XTAL(15'000'000) / 3);
+	m_iop->set_addrmap(AS_PROGRAM, &altos586_hdc_device::altos586_hdc_mem);
+	m_iop->set_addrmap(AS_IO, &altos586_hdc_device::altos586_hdc_io);
+	m_iop->set_data_width(16);
+
+	harddisk_image_device &hdd0(HARDDISK(config, "hdd0", 0));
+	hdd0.set_device_load(FUNC(altos586_hdc_device::hdd0_load));
+	hdd0.set_device_unload(FUNC(altos586_hdc_device::hdd0_unload));
+
+	harddisk_image_device &hdd1(HARDDISK(config, "hdd1", 0));
+	hdd1.set_device_load(FUNC(altos586_hdc_device::hdd1_load));
+	hdd1.set_device_unload(FUNC(altos586_hdc_device::hdd1_unload));
+}
+
+void altos586_hdc_device::device_reset()
+{
+	m_status = 0;
+	m_seek_status = 0xfc;
+
+	m_cyl_latch = 0;
+	m_cyl[m_drive] = 0;
+
+	memset(m_sector, 0, sizeof(m_sector));
+	m_secoffset = 0;
+}
+
+void altos586_hdc_device::attn_w(uint16_t data)
+{
+	m_iop->ca_w(ASSERT_LINE);
+	m_iop->ca_w(CLEAR_LINE);
+}
+
+void altos586_hdc_device::device_start()
+{
+}
```

Comment:
You should at least do basic save state registration here (register you data members for save states).

Reply:
Done, but I'm not sure it's correct -- I haven't yet used save/restore, so I need to figure out how to test it first.